File: packages/typescript/tests/transport.test.ts
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import test from "node:test";

import {
  AppServerClient,
  JsonLineDecoder,
  MessageRouter,
  encodeJsonRpcMessage,
} from "../src/transport";

type Fixture = {
  request: Record<string, unknown>;
  requestLine: string;
  chunks: string[];
  decodedMessages: Record<string, unknown>[];
  malformedLines: string[];
  router: {
    expectedResponseId: string;
    response: Record<string, unknown>;
    notification: Record<string, unknown>;
    orphanResponse: Record<string, unknown>;
  };
};

const fixture = JSON.parse(
  readFileSync(join(import.meta.dirname, "../../conformance/fixtures/transport_core.json"), "utf8"),
) as Fixture;

type ExpandedFixture = {
  framingCases: {
    chunks: string[];
    decodedMessages: Record<string, unknown>[];
    malformedLines: string[];
    pending: string;
  }[];
  router: {
    expectedResponseIds: (string | number)[];
    messages: Record<string, unknown>[];
    routes: { type: string; id?: string | number }[];
    notifications: Record<string, unknown>[];
  };
};

const expandedFixture = JSON.parse(
  readFileSync(
    join(import.meta.dirname, "../../conformance/fixtures/transport_expanded.json"),
    "utf8",
  ),
) as ExpandedFixture;

const fakeAppServer = join(import.meta.dirname, "../../conformance/fake_app_server.py");

test("encodes JSON-RPC messages as compact newline-delimited JSON", () => {
  assert.equal(encodeJsonRpcMessage(fixture.request), fixture.requestLine);
});

test("decodes complete, partial, notification, and malformed JSON lines", () => {
  const decoder = new JsonLineDecoder();
  const messages: Record<string, unknown>[] = [];
  const malformed: string[] = [];

  for (const chunk of fixture.chunks) {
    const decoded = decoder.feed(chunk);
    messages.push(...decoded.messages);
    malformed.push(...decoded.malformed.map((line) => line.raw));
  }

  assert.deepEqual(messages, fixture.decodedMessages);
  assert.deepEqual(malformed, fixture.malformedLines);
});

test("routes expected responses and preserves unknown notifications", () => {
  const router = new MessageRouter();
  router.expectResponse(fixture.router.expectedResponseId);

  assert.deepEqual(router.route(fixture.router.response), {
    type: "response",
    id: fixture.router.expectedResponseId,
    message: fixture.router.response,
  });

  assert.deepEqual(router.route(fixture.router.notification), {
    type: "notification",
    message: fixture.router.notification,
  });

  assert.deepEqual(router.route(fixture.router.orphanResponse), {
    type: "orphanResponse",
    id: "req-2",
    message: fixture.router.orphanResponse,
  });

  assert.deepEqual(router.notifications, [fixture.router.notification]);
});

test("decodes expanded shared framing cases", () => {
  for (const framingCase of expandedFixture.framingCases) {
    const decoder = new JsonLineDecoder();
    const messages: Record<string, unknown>[] = [];
    const malformed: string[] = [];

    for (const chunk of framingCase.chunks) {
      const decoded = decoder.feed(chunk);
      messages.push(...decoded.messages);
      malformed.push(...decoded.malformed.map((line) => line.raw));
    }

    assert.deepEqual(messages, framingCase.decodedMessages);
    assert.deepEqual(malformed, framingCase.malformedLines);
    assert.equal(decoder.pending, framingCase.pending);
  }
});

test("routes expanded shared message cases", () => {
  const router = new MessageRouter();
  for (const id of expandedFixture.router.expectedResponseIds) {
    router.expectResponse(id);
  }

  const routes = expandedFixture.router.messages.map((message) => router.route(message));
  assert.deepEqual(
    routes.map((route) => ({ type: route.type, id: "id" in route ? route.id : undefined })),
    expandedFixture.router.routes.map((route) => ({ type: route.type, id: route.id })),
  );
  assert.deepEqual(router.notifications, expandedFixture.router.notifications);
});

test("app-server client handles request responses and errors", async () => {
  const client = AppServerClient.start({ command: "python3", args: [fakeAppServer] });
  try {
    assert.deepEqual(await client.request("sdk/echo", { value: 42 }), { value: 42 });

    await assert.rejects(client.request("sdk/error", { retry: false }), /fake failure/);
  } finally {
    await client.close();
  }
});

test("app-server client sends and receives notifications", async () => {
  const client = AppServerClient.start({ command: "python3", args: [fakeAppServer] });
  try {
    await client.notify("sdk/client-notification", { from: "client" });
    assert.deepEqual(await client.nextEvent(), {
      type: "notification",
      message: {
        method: "fake/clientNotificationReceived",
        params: { from: "client" },
      },
    });

    assert.deepEqual(await client.request("sdk/notify-server", { from: "server" }), {
      notified: true,
    });
    assert.deepEqual(await client.nextEvent(), {
      type: "notification",
      message: {
        method: "fake/notification",
        params: { from: "server" },
      },
    });
  } finally {
    await client.close();
  }
});

test("app-server client surfaces server requests and client responses", async () => {
  const client = AppServerClient.start({ command: "python3", args: [fakeAppServer] });
  try {
    assert.deepEqual(await client.request("sdk/request-client", { question: "approve?" }), {
      requested: true,
    });
    const event = await client.nextEvent();
    assert.deepEqual(event, {
      type: "serverRequest",
      id: "server-1",
      message: {
        id: "server-1",
        method: "fake/serverRequest",
        params: { question: "approve?" },
      },
    });

    if (event.type !== "serverRequest") {
      throw new Error("expected server request");
    }
    await client.respond(event.id, { approved: true });
    assert.deepEqual(await client.nextEvent(), {
      type: "notification",
      message: {
        method: "fake/serverRequestResolved",
        params: { id: "server-1", result: { approved: true } },
      },
    });
  } finally {
    await client.close();
  }
});

test("app-server client surfaces malformed output and process exit", async () => {
  const client = AppServerClient.start({ command: "python3", args: [fakeAppServer] });
  try {
    assert.deepEqual(await client.request("sdk/malformed"), { malformed: true });
    assert.deepEqual(await client.nextEvent(), {
      type: "malformed",
      raw: "not-json",
    });

    await assert.rejects(client.request("sdk/exit"), /app-server closed/);
    assert.deepEqual(await client.nextEvent(), {
      type: "exit",
      code: 7,
      signal: null,
    });
    await client.close();
    await client.close();
  } finally {
    await client.close();
  }
});

File: packages/elixir/test/transport_test.exs
defmodule CodexSdk.TransportTest do
  use ExUnit.Case, async: true

  @fixture_path Path.expand("../../conformance/fixtures/transport_core.json", __DIR__)
  @expanded_fixture_path Path.expand(
                           "../../conformance/fixtures/transport_expanded.json",
                           __DIR__
                         )
  @fake_app_server_path Path.expand("../../conformance/fake_app_server.py", __DIR__)

  test "encodes JSON-RPC messages as compact newline-delimited JSON" do
    fixture = fixture()

    assert IO.iodata_to_binary(CodexSdk.JsonRpc.encode(fixture["request"])) ==
             fixture["requestLine"]
  end

  test "decodes complete, partial, notification, and malformed JSON lines" do
    fixture = fixture()
    state = CodexSdk.LineFraming.new()

    {state, messages, malformed} =
      Enum.reduce(fixture["chunks"], {state, [], []}, fn chunk, {state, messages, malformed} ->
        {state, events} = CodexSdk.LineFraming.feed(state, chunk)

        {
          state,
          messages ++ for({:message, message} <- events, do: message),
          malformed ++ for({:malformed, raw} <- events, do: raw)
        }
      end)

    assert state.pending == ""
    assert messages == fixture["decodedMessages"]
    assert malformed == fixture["malformedLines"]
  end

  test "routes expected responses and preserves unknown notifications" do
    fixture = fixture()
    router_fixture = fixture["router"]

    router =
      CodexSdk.MessageRouter.new()
      |> CodexSdk.MessageRouter.expect_response(router_fixture["expectedResponseId"])

    {router, routed_response} = CodexSdk.MessageRouter.route(router, router_fixture["response"])

    assert routed_response ==
             {:response, router_fixture["expectedResponseId"], router_fixture["response"]}

    {router, routed_notification} =
      CodexSdk.MessageRouter.route(router, router_fixture["notification"])

    assert routed_notification == {:notification, router_fixture["notification"]}

    {router, routed_orphan} =
      CodexSdk.MessageRouter.route(router, router_fixture["orphanResponse"])

    assert routed_orphan == {:orphan_response, "req-2", router_fixture["orphanResponse"]}
    assert router.notifications == [router_fixture["notification"]]
  end

  test "decodes expanded shared framing cases" do
    fixture = expanded_fixture()

    for framing_case <- fixture["framingCases"] do
      state = CodexSdk.LineFraming.new()

      {state, messages, malformed} =
        Enum.reduce(framing_case["chunks"], {state, [], []}, fn chunk,
                                                                {state, messages, malformed} ->
          {state, events} = CodexSdk.LineFraming.feed(state, chunk)

          {
            state,
            messages ++ for({:message, message} <- events, do: message),
            malformed ++ for({:malformed, raw} <- events, do: raw)
          }
        end)

      assert messages == framing_case["decodedMessages"]
      assert malformed == framing_case["malformedLines"]
      assert state.pending == framing_case["pending"]
    end
  end

  test "routes expanded shared message cases" do
    fixture = expanded_fixture()["router"]

    router =
      Enum.reduce(fixture["expectedResponseIds"], CodexSdk.MessageRouter.new(), fn id, router ->
        CodexSdk.MessageRouter.expect_response(router, id)
      end)

    {routes, router} =
      Enum.map_reduce(fixture["messages"], router, fn message, router ->
        {router, routed} = CodexSdk.MessageRouter.route(router, message)
        {routed, router}
      end)

    assert Enum.map(routes, &route_summary/1) == fixture["routes"]
    assert router.notifications == fixture["notifications"]
  end

  test "app-server client handles request responses and errors" do
    {:ok, server} = fake_app_server()

    try do
      assert {:ok, %{"value" => "42"}} =
               CodexSdk.AppServer.request(server, "sdk/echo", %{"value" => "42"})

      assert {:error, %CodexSdk.AppServer.JsonRpcError{message: "fake failure"}} =
               CodexSdk.AppServer.request(server, "sdk/error", %{"retry" => false})
    after
      CodexSdk.AppServer.stop(server)
    end
  end

  test "app-server client sends and receives notifications" do
    {:ok, server} = fake_app_server()

    try do
      assert :ok =
               CodexSdk.AppServer.notify(server, "sdk/client-notification", %{"from" => "client"})

      assert_receive {:usetemi_codex_sdk_app_server_event, ^server,
                      {:notification,
                       %{
                         "method" => "fake/clientNotificationReceived",
                         "params" => %{"from" => "client"}
                       }}}

      assert {:ok, %{"notified" => true}} =
               CodexSdk.AppServer.request(server, "sdk/notify-server", %{"from" => "server"})

      assert_receive {:usetemi_codex_sdk_app_server_event, ^server,
                      {:notification,
                       %{"method" => "fake/notification", "params" => %{"from" => "server"}}}}
    after
      CodexSdk.AppServer.stop(server)
    end
  end

  test "app-server client surfaces server requests and client responses" do
    {:ok, server} = fake_app_server()

    try do
      assert {:ok, %{"requested" => true}} =
               CodexSdk.AppServer.request(server, "sdk/request-client", %{
                 "question" => "approve?"
               })

      assert_receive {:usetemi_codex_sdk_app_server_event, ^server,
                      {:server_request, "server-1",
                       %{
                         "id" => "server-1",
                         "method" => "fake/serverRequest",
                         "params" => %{"question" => "approve?"}
                       }}}

      assert :ok = CodexSdk.AppServer.respond(server, "server-1", %{"approved" => true})

      assert_receive {:usetemi_codex_sdk_app_server_event, ^server,
                      {:notification,
                       %{
                         "method" => "fake/serverRequestResolved",
                         "params" => %{
                           "id" => "server-1",
                           "result" => %{"approved" => true}
                         }
                       }}}
    after
      CodexSdk.AppServer.stop(server)
    end
  end

  test "app-server client surfaces malformed output and process exit" do
    {:ok, server} = fake_app_server()

    assert {:ok, %{"malformed" => true}} = CodexSdk.AppServer.request(server, "sdk/malformed")
    assert_receive {:usetemi_codex_sdk_app_server_event, ^server, {:malformed, "not-json"}}

    assert {:error, %CodexSdk.AppServer.ClosedError{code: 7}} =
             CodexSdk.AppServer.request(server, "sdk/exit")

    assert_receive {:usetemi_codex_sdk_app_server_event, ^server, {:exit, 7}}
    assert :ok = CodexSdk.AppServer.stop(server)
    assert :ok = CodexSdk.AppServer.stop(server)
  end

  defp fixture do
    @fixture_path
    |> File.read!()
    |> Jason.decode!()
  end

  defp expanded_fixture do
    @expanded_fixture_path
    |> File.read!()
    |> Jason.decode!()
  end

  defp fake_app_server do
    CodexSdk.AppServer.start_link(
      command: "python3",
      args: [@fake_app_server_path],
      owner: self()
    )
  end

  defp route_summary({:response, id, _message}), do: %{"type" => "response", "id" => id}

  defp route_summary({:error_response, id, _message}),
    do: %{"type" => "errorResponse", "id" => id}

  defp route_summary({:server_request, id, _message}),
    do: %{"type" => "serverRequest", "id" => id}

  defp route_summary({:orphan_response, id, _message}),
    do: %{"type" => "orphanResponse", "id" => id}

  defp route_summary({:notification, _message}), do: %{"type" => "notification"}
  defp route_summary({:unknown, _message}), do: %{"type" => "unknown"}
end
